# VWEHX vs CPI: Full EDA Suite

This notebook builds an end-to-end exploratory analysis pipeline to detect patterns between the VWEHX (Vanguard High-Yield Corporate Inv) and Hong Kong CPI series.

## What this covers
- Data loading and robust cleaning
- CPI parsing from a multi-row header file
- Daily-to-monthly VWEHX feature engineering
- Temporal alignment and merge
- Trend and distribution visualizations
- Correlation, lagged correlation, and rolling correlation
- Regime and seasonality checks
- OLS regression diagnostics
- Granger causality tests


In [ ]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

from scipy import stats
from statsmodels.tsa.stattools import adfuller, grangercausalitytests
import statsmodels.api as sm

import yfinance as yf
from datetime import datetime

# DATA_VWEHX = 'vwehx_data.csv'
DATA_CPI = 'HK_CPI_data.csv'


## 0) Fetch S&P Data from yfinance 

In [37]:
start_date = "2000-01-01"
vwehx_raw = yf.download('VWEHX', start=start_date, end=datetime.today().date(), progress=False, auto_adjust=True) # Prices are adjusted for dividends
vwehx_raw.reset_index(inplace=True)
vwehx_raw['Date'] = vwehx_raw['Date'].dt.strftime('%Y-%m-%d')
vwehx_info = yf.Ticker('VWEHX')

In [38]:
vwehx_info.info

{'address1': 'Valley Forge PA 19482',
 'address2': 'PO Box 2600 V26',
 'address3': 'Vanguard Fixed Income Securities Funds',
 'phone': '800-523-1036',
 'longBusinessSummary': 'The fund invests primarily in a diversified group of high-yielding, higher-risk corporate bonds-commonly known as "junk bonds"-with medium- and lower-range credit-quality ratings. It invests at least 80% of its assets in corporate bonds that are rated below Baa by Moody\'s Investors Service, Inc. (Moody\'s); have an equivalent rating by any other independent bond-rating agency; or, if unrated, are determined to be of comparable quality by the fund\'s advisor. The fund\'s high-yield bonds and loans mostly have short- and intermediate-term maturities.',
 'companyOfficers': [],
 'executiveTeam': [],
 'maxAge': 86400,
 'priceHint': 2,
 'previousClose': 5.5299997,
 'regularMarketPreviousClose': 5.5299997,
 'averageVolume': 0,
 'averageVolume10days': 0,
 'averageDailyVolume10Day': 0,
 'yield': 0.0618,
 'ytdReturn': 0.1

In [39]:
vwehx_raw.head()

Price,Date,Close,High,Low,Open,Volume
Ticker,,VWEHX,VWEHX,VWEHX,VWEHX,VWEHX
0,2000-01-03,1.327131,1.327131,1.327131,1.327131,0
1,2000-01-04,1.323528,1.323528,1.323528,1.323528,0
2,2000-01-05,1.319928,1.319928,1.319928,1.319928,0
3,2000-01-06,1.319928,1.319928,1.319928,1.319928,0
4,2000-01-07,1.319928,1.319928,1.319928,1.319928,0


In [40]:
vwehx_raw.columns = vwehx_raw.columns.get_level_values(0)
vwehx_raw.columns.name = None
vwehx_raw.head()

,Date,Close,High,Low,Open,Volume
0,2000-01-03,1.327131,1.327131,1.327131,1.327131,0
1,2000-01-04,1.323528,1.323528,1.323528,1.323528,0
2,2000-01-05,1.319928,1.319928,1.319928,1.319928,0
3,2000-01-06,1.319928,1.319928,1.319928,1.319928,0
4,2000-01-07,1.319928,1.319928,1.319928,1.319928,0


## 1) Load raw data

In [41]:
cpi_raw = pd.read_csv(DATA_CPI, header=None)

print('VWEHX raw shape:', vwehx_raw.shape)
print('CPI raw shape:', cpi_raw.shape)

display(vwehx_raw.head())
display(vwehx_raw.head(10))


VWEHX raw shape: (6581, 6)
CPI raw shape: (693, 14)


,Date,Close,High,Low,Open,Volume
0,2000-01-03,1.327131,1.327131,1.327131,1.327131,0
1,2000-01-04,1.323528,1.323528,1.323528,1.323528,0
2,2000-01-05,1.319928,1.319928,1.319928,1.319928,0
3,2000-01-06,1.319928,1.319928,1.319928,1.319928,0
4,2000-01-07,1.319928,1.319928,1.319928,1.319928,0


,Date,Close,High,Low,Open,Volume
0,2000-01-03,1.327131,1.327131,1.327131,1.327131,0
1,2000-01-04,1.323528,1.323528,1.323528,1.323528,0
2,2000-01-05,1.319928,1.319928,1.319928,1.319928,0
3,2000-01-06,1.319928,1.319928,1.319928,1.319928,0
4,2000-01-07,1.319928,1.319928,1.319928,1.319928,0
5,2000-01-10,1.318127,1.318127,1.318127,1.318127,0
6,2000-01-11,1.316324,1.316324,1.316324,1.316324,0
7,2000-01-12,1.314525,1.314525,1.314525,1.314525,0
8,2000-01-13,1.312724,1.312724,1.312724,1.312724,0
9,2000-01-14,1.312724,1.312724,1.312724,1.312724,0


## 2) Clean CPI table (multi-row header)

The CPI file includes title/meta rows and grouped headers. We detect the true `Year, Month` row and rebuild readable column names.


In [42]:
raw = cpi_raw.copy()

#--------------------------- cleaning the file ----------------------------

# Locate the true header row by finding where col0 == "Year" and col1 == "Month".
# This is more robust than hard-coding a row number because source files may shift.
hdr_idx = raw.index[
    (raw.iloc[:, 0].astype(str).str.strip().eq("Year"))
    & (raw.iloc[:, 1].astype(str).str.strip().eq("Month"))
][0]

# The table uses a 2-row grouped header above the "Year/Month" row:
# - h1: high-level group (e.g., CPI type)
# - h2: metric (e.g., Index, YoY change, MoM change)
h1 = raw.iloc[hdr_idx - 2].ffill()  # forward-fill group names across merged-like cells
h2 = raw.iloc[hdr_idx - 1]

# Build clean, unique column names by combining h1 and h2.
new_cols = []
seen = {}  # tracks duplicates to enforce uniqueness

for i in range(raw.shape[1]):
    if i == 0:
        col = "Year"
    elif i == 1:
        col = "Month"
    else:
        a = str(h1[i]).strip() if pd.notna(h1[i]) else ""
        b = str(h2[i]).strip() if pd.notna(h2[i]) else ""

        # Prefer "Group - Metric"; fall back to whichever exists; else generated name.
        col = f"{a} - {b}" if a and b else (a or b or f"col_{i}")

        # Normalize symbols/text to avoid awkward column names in downstream code.
        col = col.replace("%", "pct")

        # Remove non-breaking spaces and collapse extra whitespace to prevent hidden
        # near-duplicates (e.g., visually same but technically different names).
        col = col.replace("\xa0", " ")
        col = " ".join(col.split())

    # Ensure uniqueness in case normalization produces duplicate names.
    if col in seen:
        seen[col] += 1
        col = f"{col}_{seen[col]}"
    else:
        seen[col] = 0

    new_cols.append(col)

# Extract data rows only (everything below the header marker row) and assign new columns.
cpi_df = raw.iloc[hdr_idx + 1 :].copy()
cpi_df.columns = new_cols

# Clean and convert all value columns (exclude Year/Month at positions 0 and 1).
for col in cpi_df.columns[2:]:
    cpi_df[col] = (
        cpi_df[col]
        .astype(str)
        # Remove bracketed footnotes like "[1]" that break numeric parsing.
        .str.replace(r"\s*\[.*?\]", "", regex=True)
        # Standardize missing-value markers before numeric conversion.
        .replace({"N.A.": pd.NA, "nan": pd.NA, "": pd.NA})
    )
    # Coerce invalid strings to NaN so numeric operations work safely.
    cpi_df[col] = pd.to_numeric(cpi_df[col], errors="coerce")

# Convert Year/Month to year and month numbers only (nullable integers).
year_dt = pd.to_datetime(
    cpi_df["Year"].astype("string").str.strip(),
    format="%Y",
    errors="coerce",
)

month_str = cpi_df["Month"].astype("string").str.strip()
month_num = pd.to_numeric(month_str, errors="coerce")
month_num = month_num.fillna(pd.to_datetime(month_str, format="%b", errors="coerce").dt.month)
month_num = month_num.fillna(pd.to_datetime(month_str, format="%B", errors="coerce").dt.month)

cpi_df["Year"] = year_dt.dt.year.astype("Int64")
cpi_df["Month"] = month_num.astype("Int64")

# Keep only rows with a valid year (drops footer/notes rows).
cpi_df = cpi_df[cpi_df["Year"].notna()].reset_index(drop=True)

# Show only Year and Month preview.
display(cpi_df[["Year", "Month"]].head())

# Quick sanity check of the cleaned output.
display(cpi_df.head())

# Shorten cpi_df column names
short_name_map = {
    "Consumer Price Index": "CPI",
    "Composite CPI": "CCPI",
    "CPI (A)": "CPI_A",
    "CPI (B)": "CPI_B",
    "CPI (C)": "CPI_C",
    "Year-on-year pct change": "yoy_pct",
    "Month-to-month pct change": "mom_pct",
    "Index": "idx",
}

cpi_df = cpi_df.rename(
    columns=lambda col: col if col in ["Year", "Month"] else (
        col.replace("Consumer Price Index", short_name_map["Consumer Price Index"])
           .replace("Composite CPI", short_name_map["Composite CPI"])
           .replace("CPI (A)", short_name_map["CPI (A)"])
           .replace("CPI (B)", short_name_map["CPI (B)"])
           .replace("CPI (C)", short_name_map["CPI (C)"])
           .replace("Year-on-year pct change", short_name_map["Year-on-year pct change"])
           .replace("Month-to-month pct change", short_name_map["Month-to-month pct change"])
           .replace("Index", short_name_map["Index"])
           .replace(" - ", "_")
           .replace(" ", "")
    )
)

display(cpi_df.columns)

# Use monthly records only (annual rows have Month = <NA>)
monthly = cpi_df[cpi_df["Month"].notna()].copy()
monthly["Date"] = pd.to_datetime(
    dict(year=monthly["Year"].astype(int), month=monthly["Month"].astype(int), day=1)
) + pd.offsets.MonthEnd(0)
monthly = monthly.sort_values("Date").reset_index(drop=True)

# Keep downstream variable name unchanged for the rest of the notebook
cpi = monthly.copy()
print('Clean CPI monthly shape:', cpi.shape)
print('Date range:', cpi['Date'].min().date(), '->', cpi['Date'].max().date())
display(cpi.head())

,Year,Month
0,1975,<NA>
1,1976,<NA>
2,1977,<NA>
3,1978,<NA>
4,1979,<NA>


,Year,Month,Composite Consumer Price Index - Index,Composite Consumer Price Index - Year-on-year pct change,Composite Consumer Price Index - Month-to-month pct change,Consumer Price Index (A) - Index,Consumer Price Index (A) - Year-on-year pct change,Consumer Price Index (A) - Month-to-month pct change,Consumer Price Index (B) - Index,Consumer Price Index (B) - Year-on-year pct change,Consumer Price Index (B) - Month-to-month pct change,Consumer Price Index (C) - Index,Consumer Price Index (C) - Year-on-year pct change,Consumer Price Index (C) - Month-to-month pct change
0,1975,<NA>,NaN,NaN,NaN,12.7,NaN,NaN,13.2,NaN,NaN,11.9,NaN,NaN
1,1976,<NA>,NaN,NaN,NaN,13.1,3.4,NaN,13.7,4.0,NaN,12.5,4.4,NaN
2,1977,<NA>,NaN,NaN,NaN,13.9,5.8,NaN,14.4,5.5,NaN,13.1,5.0,NaN
3,1978,<NA>,NaN,NaN,NaN,14.7,5.9,NaN,15.3,5.9,NaN,13.8,5.8,NaN
4,1979,<NA>,NaN,NaN,NaN,16.4,11.6,NaN,17.0,11.5,NaN,15.5,12.4,NaN


Index(['Year', 'Month', 'CCPI_idx', 'CCPI_yoy_pct', 'CCPI_mom_pct',
       'CPI_A_idx', 'CPI_A_yoy_pct', 'CPI_A_mom_pct', 'CPI_B_idx',
       'CPI_B_yoy_pct', 'CPI_B_mom_pct', 'CPI_C_idx', 'CPI_C_yoy_pct',
       'CPI_C_mom_pct'],
      dtype='str')

Clean CPI monthly shape: (618, 15)
Date range: 1974-07-31 -> 2025-12-31


,Year,Month,CCPI_idx,CCPI_yoy_pct,CCPI_mom_pct,CPI_A_idx,CPI_A_yoy_pct,CPI_A_mom_pct,CPI_B_idx,CPI_B_yoy_pct,CPI_B_mom_pct,CPI_C_idx,CPI_C_yoy_pct,CPI_C_mom_pct,Date
0,1974,7,NaN,NaN,NaN,12.6,NaN,NaN,13.0,NaN,NaN,11.8,NaN,NaN,1974-07-31
1,1974,8,NaN,NaN,NaN,12.5,NaN,-1.3,12.9,NaN,-0.8,11.8,NaN,-0.3,1974-08-31
2,1974,9,NaN,NaN,NaN,12.5,NaN,0.3,13.0,NaN,0.2,11.8,NaN,0.3,1974-09-30
3,1974,10,NaN,NaN,NaN,12.8,NaN,2.0,13.2,NaN,1.6,11.9,NaN,0.6,1974-10-31
4,1974,11,NaN,NaN,NaN,12.8,NaN,0.5,13.2,NaN,0.4,11.9,NaN,-0.1,1974-11-30


## 3) Clean VWEHX and create monthly features

In [43]:
vwehx = vwehx_raw.copy()

# Parse date and numeric fields
vwehx['Date'] = pd.to_datetime(vwehx['Date'], errors='coerce')
for c in ['Close', 'High', 'Low', 'Open', 'Volume']:
    vwehx[c] = pd.to_numeric(vwehx[c], errors='coerce')

vwehx = vwehx.dropna(subset=['Date', 'Close']).sort_values('Date').reset_index(drop=True)

# Explicit filter requested: 2000 onwards
vwehx = vwehx[vwehx['Date'] >= pd.Timestamp('2000-01-01')].copy()

# Daily return and volatility proxy
vwehx['Daily_Return'] = vwehx['Close'].pct_change()
vwehx['Log_Return'] = np.log(vwehx['Close'] / vwehx['Close'].shift(1))

# Monthly aggregation
monthly_vwehx = (
    vwehx.set_index('Date')
       .resample('ME')
       .agg(
           VWEHX_Close=('Close', 'last'),
           VWEHX_Open=('Open', 'first'),
           VWEHX_High=('High', 'max'),
           VWEHX_Low=('Low', 'min'),
           VWEHX_DailyRet_Mean=('Daily_Return', 'mean'),
           VWEHX_DailyRet_Std=('Daily_Return', 'std'),
           VWEHX_LogRet_Sum=('Log_Return', 'sum')
       )
       .reset_index()
)

monthly_vwehx['VWEHX_Monthly_Return'] = monthly_vwehx['VWEHX_Close'].pct_change()
monthly_vwehx['VWEHX_YoY_Return'] = monthly_vwehx['VWEHX_Close'].pct_change(12)

print('Monthly VWEHX shape:', monthly_vwehx.shape)
print('Date range:', monthly_vwehx['Date'].min().date(), '->', monthly_vwehx['Date'].max().date())
display(monthly_vwehx.head())


Monthly VWEHX shape: (315, 10)
Date range: 2000-01-31 -> 2026-03-31


,Date,VWEHX_Close,VWEHX_Open,VWEHX_High,VWEHX_Low,VWEHX_DailyRet_Mean,VWEHX_DailyRet_Std,VWEHX_LogRet_Sum,VWEHX_Monthly_Return,VWEHX_YoY_Return
0,2000-01-31,1.320538,1.327131,1.327131,1.309123,-0.000260,0.002211,-0.004981,NaN,NaN
1,2000-02-29,1.322402,1.320538,1.322402,1.315096,0.000072,0.001706,0.001411,0.001412,NaN
2,2000-03-31,1.310408,1.322402,1.322402,1.302310,-0.000395,0.001758,-0.009111,-0.009070,NaN
3,2000-04-30,1.320236,1.312250,1.320236,1.308567,0.000396,0.002362,0.007472,0.007500,NaN
4,2000-05-31,1.304366,1.320236,1.320236,1.294276,-0.000547,0.002121,-0.012093,-0.012020,NaN


## 4) Select CPI features and merge with VWEHX

In [44]:
cpi.columns.tolist()

['Year',
 'Month',
 'CCPI_idx',
 'CCPI_yoy_pct',
 'CCPI_mom_pct',
 'CPI_A_idx',
 'CPI_A_yoy_pct',
 'CPI_A_mom_pct',
 'CPI_B_idx',
 'CPI_B_yoy_pct',
 'CPI_B_mom_pct',
 'CPI_C_idx',
 'CPI_C_yoy_pct',
 'CPI_C_mom_pct',
 'Date']

In [45]:
# Pick commonly used CPI signals (supports both long source names and shortened names)
def pick_col(columns, all_keywords=None, any_keywords=None):
    all_keywords = all_keywords or []
    any_keywords = any_keywords or []
    for col in columns:
        col_l = col.lower()
        if all(k.lower() in col_l for k in all_keywords):
            if not any_keywords or any(k.lower() in col_l for k in any_keywords):
                return col
    return None


cpi_cols = cpi.columns.tolist()
cpi_groups = ['CCPI', 'CPI_A', 'CPI_B', 'CPI_C']

selected = ['Date']
rename_map = {}

for grp in cpi_groups:
    idx_col = pick_col(cpi_cols, all_keywords=[grp], any_keywords=['idx', 'index'])
    yoy_col = pick_col(cpi_cols, all_keywords=[grp], any_keywords=['yoy_pct', 'year-on-year'])
    mom_col = pick_col(cpi_cols, all_keywords=[grp], any_keywords=['mom_pct', 'month-to-month'])

    if grp == 'CCPI':
        idx_col = idx_col or pick_col(cpi_cols, all_keywords=['composite consumer price index'], any_keywords=['index'])
        yoy_col = yoy_col or pick_col(cpi_cols, all_keywords=['composite consumer price index'], any_keywords=['year-on-year'])
        mom_col = mom_col or pick_col(cpi_cols, all_keywords=['composite consumer price index'], any_keywords=['month-to-month'])

    if idx_col is not None:
        selected.append(idx_col)
        rename_map[idx_col] = f'{grp}_Index'
    if yoy_col is not None:
        selected.append(yoy_col)
        rename_map[yoy_col] = f'{grp}_YoY_Pct'
    if mom_col is not None:
        selected.append(mom_col)
        rename_map[mom_col] = f'{grp}_MoM_Pct'

selected = ['Date'] + list(dict.fromkeys([c for c in selected if c != 'Date']))
cpi_sel = cpi[selected].copy().rename(columns=rename_map)

# Build missing YoY/MoM from index if needed
for grp in cpi_groups:
    idx_name = f'{grp}_Index'
    yoy_name = f'{grp}_YoY_Pct'
    mom_name = f'{grp}_MoM_Pct'
    if idx_name in cpi_sel.columns and yoy_name not in cpi_sel.columns:
        cpi_sel[yoy_name] = cpi_sel[idx_name].pct_change(12) * 100
    if idx_name in cpi_sel.columns and mom_name not in cpi_sel.columns:
        cpi_sel[mom_name] = cpi_sel[idx_name].pct_change(1) * 100

# Backward-compatible aliases used by older cells
if 'CCPI_Index' in cpi_sel.columns and 'CPI_Index' not in cpi_sel.columns:
    cpi_sel['CPI_Index'] = cpi_sel['CCPI_Index']
if 'CCPI_YoY_Pct' in cpi_sel.columns and 'CPI_YoY_Pct' not in cpi_sel.columns:
    cpi_sel['CPI_YoY_Pct'] = cpi_sel['CCPI_YoY_Pct']
if 'CCPI_MoM_Pct' in cpi_sel.columns and 'CPI_MoM_Pct' not in cpi_sel.columns:
    cpi_sel['CPI_MoM_Pct'] = cpi_sel['CCPI_MoM_Pct']

df = monthly_vwehx.merge(cpi_sel, on='Date', how='inner').sort_values('Date').reset_index(drop=True)

# Final guard: analysis period starts from 2000-01-01
df = df[df['Date'] >= pd.Timestamp('2000-01-01')].reset_index(drop=True)

print('Merged dataset shape:', df.shape)
print('Merged date range:', df['Date'].min().date(), '->', df['Date'].max().date())
print('CPI series available:', [c for c in df.columns if c.endswith('_Index') or c.endswith('_YoY_Pct') or c.endswith('_MoM_Pct')])
display(df.head())
display(df.tail())

Merged dataset shape: (312, 25)
Merged date range: 2000-01-31 -> 2025-12-31
CPI series available: ['CCPI_Index', 'CCPI_YoY_Pct', 'CCPI_MoM_Pct', 'CPI_A_Index', 'CPI_A_YoY_Pct', 'CPI_A_MoM_Pct', 'CPI_B_Index', 'CPI_B_YoY_Pct', 'CPI_B_MoM_Pct', 'CPI_C_Index', 'CPI_C_YoY_Pct', 'CPI_C_MoM_Pct', 'CPI_Index', 'CPI_YoY_Pct', 'CPI_MoM_Pct']


,Date,VWEHX_Close,VWEHX_Open,VWEHX_High,VWEHX_Low,VWEHX_DailyRet_Mean,VWEHX_DailyRet_Std,VWEHX_LogRet_Sum,VWEHX_Monthly_Return,VWEHX_YoY_Return,...,CPI_A_MoM_Pct,CPI_B_Index,CPI_B_YoY_Pct,CPI_B_MoM_Pct,CPI_C_Index,CPI_C_YoY_Pct,CPI_C_MoM_Pct,CPI_Index,CPI_YoY_Pct,CPI_MoM_Pct
0,2000-01-31,1.320538,1.327131,1.327131,1.309123,-0.000260,0.002211,-0.004981,NaN,NaN,...,-0.5,71.8,-5.7,-0.6,72.9,-6.2,-1.0,71.5,-5.3,-0.7
1,2000-02-29,1.322402,1.320538,1.322402,1.315096,0.000072,0.001706,0.001411,0.001412,NaN,...,0.0,71.7,-5.4,-0.1,72.9,-5.8,0.0,71.4,-5.1,-0.1
2,2000-03-31,1.310408,1.322402,1.322402,1.302310,-0.000395,0.001758,-0.009111,-0.009070,NaN,...,-0.5,71.4,-5.3,-0.4,72.7,-5.7,-0.3,71.2,-5.0,-0.4
3,2000-04-30,1.320236,1.312250,1.320236,1.308567,0.000396,0.002362,0.007472,0.007500,NaN,...,0.0,71.4,-4.6,0.0,72.9,-5.2,0.3,71.2,-4.4,0.1
4,2000-05-31,1.304366,1.320236,1.320236,1.294276,-0.000547,0.002121,-0.012093,-0.012020,NaN,...,-0.2,71.2,-4.6,-0.3,72.8,-5.6,-0.2,71.1,-4.5,-0.2


,Date,VWEHX_Close,VWEHX_Open,VWEHX_High,VWEHX_Low,VWEHX_DailyRet_Mean,VWEHX_DailyRet_Std,VWEHX_LogRet_Sum,VWEHX_Monthly_Return,VWEHX_YoY_Return,...,CPI_A_MoM_Pct,CPI_B_Index,CPI_B_YoY_Pct,CPI_B_MoM_Pct,CPI_C_Index,CPI_C_YoY_Pct,CPI_C_MoM_Pct,CPI_Index,CPI_YoY_Pct,CPI_MoM_Pct
307,2025-08-31,5.409595,5.342545,5.409595,5.342545,0.000595,0.001523,0.012472,0.012550,0.077846,...,0.1,107.7,1.0,0.1,107.5,0.9,0.2,109.0,1.1,0.1
308,2025-09-30,5.447756,5.390101,5.447756,5.390101,0.000336,0.001834,0.007030,0.007054,0.072128,...,0.1,107.8,0.9,0.1,107.5,0.8,0.0,109.1,1.1,0.1
309,2025-10-31,5.466523,5.447756,5.466523,5.408564,0.000151,0.001935,0.003439,0.003445,0.084051,...,0.1,108.1,1.1,0.3,108.0,1.0,0.4,109.4,1.2,0.3
310,2025-11-30,5.504141,5.456673,5.504141,5.427125,0.000363,0.002161,0.006858,0.006882,0.079979,...,0.0,108.1,1.1,0.0,108.0,1.1,0.0,109.4,1.2,0.0
311,2025-12-31,5.541899,5.494242,5.541899,5.494242,0.000312,0.001373,0.006837,0.006860,0.093746,...,0.1,108.5,1.3,0.3,108.6,1.4,0.6,109.8,1.4,0.3


## 5) Data quality snapshot

In [46]:
summary = pd.DataFrame({
    'dtype': df.dtypes.astype(str),
    'missing': df.isna().sum(),
    'missing_pct': (df.isna().mean() * 100).round(2),
    'n_unique': df.nunique()
})

display(summary)
display(df.describe(include='all').T)


,dtype,missing,missing_pct,n_unique
Date,datetime64[us],0,0.00,312
VWEHX_Close,float64,0,0.00,312
VWEHX_Open,float64,0,0.00,312
VWEHX_High,float64,0,0.00,286
VWEHX_Low,float64,0,0.00,312
VWEHX_DailyRet_Mean,float64,0,0.00,312
VWEHX_DailyRet_Std,float64,0,0.00,312
VWEHX_LogRet_Sum,float64,0,0.00,312
VWEHX_Monthly_Return,float64,1,0.32,311
VWEHX_YoY_Return,float64,12,3.85,300


,count,mean,min,25%,50%,75%,max,std
Date,312,2013-01-14 03:55:23.076923,2000-01-31 00:00:00,2006-07-23 06:00:00,2013-01-15 12:00:00,2019-07-07 18:00:00,2025-12-31 00:00:00,NaN
VWEHX_Close,312.0,2.942122,1.285519,1.776537,2.951026,4.01626,5.541899,1.238026
VWEHX_Open,312.0,2.929188,1.283562,1.765773,2.936385,3.998641,5.494242,1.233049
VWEHX_High,312.0,2.959472,1.314813,1.784673,2.969311,4.07068,5.541899,1.242756
VWEHX_Low,312.0,2.899814,1.274212,1.755292,2.916751,3.974059,5.494242,1.224
VWEHX_DailyRet_Mean,312.0,0.000224,-0.00719,-0.000196,0.000323,0.000675,0.00373,0.001044
VWEHX_DailyRet_Std,312.0,0.002522,0.000749,0.001591,0.002063,0.002765,0.018872,0.001744
VWEHX_LogRet_Sum,312.0,0.004581,-0.167218,-0.00436,0.006697,0.014302,0.081414,0.022186
VWEHX_Monthly_Return,311.0,0.004866,-0.153985,-0.004112,0.00672,0.014429,0.08482,0.021904
VWEHX_YoY_Return,300.0,0.060391,-0.270453,0.020259,0.062998,0.096386,0.469183,0.083697


## 6) Trend visualization (level + standardized)

Standardized plots help compare co-movement despite different scales.


In [47]:
cpi_index_cols = [c for c in ['CCPI_Index', 'CPI_A_Index', 'CPI_B_Index', 'CPI_C_Index'] if c in df.columns]

fig = go.Figure()
fig.add_trace(go.Scatter(x=df['Date'], y=df['VWEHX_Close'], mode='lines', name='VWEHX_Close', yaxis='y1'))
for col in cpi_index_cols:
    fig.add_trace(go.Scatter(x=df['Date'], y=df[col], mode='lines', name=col, yaxis='y2'))

fig.update_layout(
    title='VWEHX vs CPI Indices (Levels)',
    xaxis_title='Date',
    yaxis=dict(title='VWEHX Close'),
    yaxis2=dict(title='CPI Index', overlaying='y', side='right'),
    hovermode='x unified',
    template='plotly_white',
    dragmode='pan',
)
fig.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})

z_cols = ['VWEHX_Close'] + cpi_index_cols
z = df[['Date'] + z_cols].copy()
for col in z_cols:
    z[col] = (z[col] - z[col].mean()) / z[col].std()

z_long = z.melt(id_vars='Date', var_name='Series', value_name='zscore')
fig2 = px.line(
    z_long,
    x='Date',
    y='zscore',
    color='Series',
    title='Standardized Co-movement: VWEHX and CPI Indices',
)
fig2.update_layout(template='plotly_white', yaxis_title='z-score', dragmode='pan')
fig2.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})


## 7) Contemporaneous relationships

In [48]:
cpi_metric_cols = [
    c for c in df.columns
    if c.endswith('_Index') or c.endswith('_YoY_Pct') or c.endswith('_MoM_Pct')
]

corr_cols = [c for c in [
    'VWEHX_Close', 'VWEHX_Monthly_Return', 'VWEHX_YoY_Return', 'VWEHX_DailyRet_Std',
    *cpi_metric_cols,
] if c in df.columns]

corr_matrix = df[corr_cols].corr()
display(corr_matrix)

fig = px.imshow(
    corr_matrix,
    color_continuous_scale='RdBu_r',
    zmin=-1,
    zmax=1,
    title='Correlation Matrix (VWEHX and CPI Variants)',
)
fig.update_layout(template='plotly_white', dragmode='pan')
fig.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})

yoy_cols = [c for c in ['CCPI_YoY_Pct', 'CPI_A_YoY_Pct', 'CPI_B_YoY_Pct', 'CPI_C_YoY_Pct'] if c in df.columns]
if 'VWEHX_Monthly_Return' in df.columns and yoy_cols:
    scatter_df = df[['VWEHX_Monthly_Return'] + yoy_cols].dropna().melt(
        id_vars='VWEHX_Monthly_Return',
        var_name='CPI_Series',
        value_name='CPI_YoY_Pct',
    )
    fig2 = px.scatter(
        scatter_df,
        x='CPI_YoY_Pct',
        y='VWEHX_Monthly_Return',
        color='CPI_Series',
        title='VWEHX Monthly Return vs CPI YoY % (All CPI Variants)',
        trendline='ols',
    )
    fig2.update_layout(template='plotly_white', dragmode='pan')
    fig2.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})
else:
    print('No YoY CPI columns available for scatter comparison.')

,VWEHX_Close,VWEHX_Monthly_Return,VWEHX_YoY_Return,VWEHX_DailyRet_Std,CCPI_Index,CCPI_YoY_Pct,CCPI_MoM_Pct,CPI_A_Index,CPI_A_YoY_Pct,CPI_A_MoM_Pct,CPI_B_Index,CPI_B_YoY_Pct,CPI_B_MoM_Pct,CPI_C_Index,CPI_C_YoY_Pct,CPI_C_MoM_Pct,CPI_Index,CPI_YoY_Pct,CPI_MoM_Pct
VWEHX_Close,1.000000,0.019204,0.030819,-0.040925,0.979899,0.363463,0.077881,0.978007,0.362005,0.038604,0.979140,0.341080,0.121408,0.978710,0.335300,0.137483,0.979899,0.363463,0.077881
VWEHX_Monthly_Return,0.019204,1.000000,0.222228,-0.365320,-0.012144,-0.033133,-0.001956,-0.010007,-0.012779,-0.000383,-0.012605,-0.035461,-0.000005,-0.013985,-0.053744,-0.014800,-0.012144,-0.033133,-0.001956
VWEHX_YoY_Return,0.030819,0.222228,1.000000,-0.471327,-0.066170,0.027613,0.053294,-0.065376,0.102105,0.026732,-0.066253,0.006812,0.094805,-0.066418,-0.056654,0.099505,-0.066170,0.027613,0.053294
VWEHX_DailyRet_Std,-0.040925,-0.365320,-0.471327,1.000000,0.044303,0.003791,0.005574,0.040369,-0.046615,0.008687,0.045288,0.019037,-0.000146,0.048030,0.050980,-0.029380,0.044303,0.003791,0.005574
CCPI_Index,0.979899,-0.012144,-0.066170,0.044303,1.000000,0.324295,0.083699,0.997762,0.335601,0.053621,0.999551,0.298445,0.112004,0.998724,0.289670,0.111749,1.000000,0.324295,0.083699
CCPI_YoY_Pct,0.363463,-0.033133,0.027613,0.003791,0.324295,1.000000,0.299597,0.302845,0.928218,0.170339,0.332610,0.986401,0.442897,0.339453,0.955655,0.432584,0.324295,1.000000,0.299597
CCPI_MoM_Pct,0.077881,-0.001956,0.053294,0.005574,0.083699,0.299597,1.000000,0.102814,0.385318,0.949338,0.074284,0.245975,0.868288,0.070519,0.206734,0.469343,0.083699,0.299597,1.000000
CPI_A_Index,0.978007,-0.010007,-0.065376,0.040369,0.997762,0.302845,0.102814,1.000000,0.324212,0.077769,0.995394,0.272996,0.118739,0.993208,0.263009,0.103537,0.997762,0.302845,0.102814
CPI_A_YoY_Pct,0.362005,-0.012779,0.102105,-0.046615,0.335601,0.928218,0.385318,0.324212,1.000000,0.279470,0.339874,0.857502,0.483211,0.342861,0.780603,0.389195,0.335601,0.928218,0.385318
CPI_A_MoM_Pct,0.038604,-0.000383,0.026732,0.008687,0.053621,0.170339,0.949338,0.077769,0.279470,1.000000,0.042204,0.112329,0.678067,0.036402,0.072321,0.178382,0.053621,0.170339,0.949338


## 8) Lead-lag analysis

Positive lag means CPI leads VWEHX by that many months.


In [49]:
def lagged_corr(x: pd.Series, y: pd.Series, lags=range(-12, 13)) -> pd.DataFrame:
    rows = []
    for lag in lags:
        rows.append({'lag': lag, 'corr': x.corr(y.shift(lag))})
    return pd.DataFrame(rows)

cpi_yoy_cols = [c for c in ['CCPI_YoY_Pct', 'CPI_A_YoY_Pct', 'CPI_B_YoY_Pct', 'CPI_C_YoY_Pct'] if c in df.columns]
cpi_mom_cols = [c for c in ['CCPI_MoM_Pct', 'CPI_A_MoM_Pct', 'CPI_B_MoM_Pct', 'CPI_C_MoM_Pct'] if c in df.columns]

pairs = []
for c_col in cpi_yoy_cols:
    if {'VWEHX_Monthly_Return', c_col}.issubset(df.columns):
        pairs.append(('VWEHX_Monthly_Return', c_col))
    if {'VWEHX_YoY_Return', c_col}.issubset(df.columns):
        pairs.append(('VWEHX_YoY_Return', c_col))
for c_col in cpi_mom_cols:
    if {'VWEHX_Monthly_Return', c_col}.issubset(df.columns):
        pairs.append(('VWEHX_Monthly_Return', c_col))

for h_col, c_col in pairs:
    lc = lagged_corr(df[h_col], df[c_col])
    valid = lc.dropna(subset=['corr']).copy()

    if valid.empty:
        print(f'{h_col} vs {c_col}: no valid lagged correlations (insufficient overlapping data).')
        continue

    best_idx = valid['corr'].abs().idxmax()
    best = valid.loc[best_idx]
    print(f'{h_col} vs {c_col}: strongest |corr| at lag={int(best["lag"])} -> corr={best["corr"]:.3f}')

    fig = px.bar(
        lc,
        x='lag',
        y='corr',
        title=f'Lagged Correlation: {h_col} vs {c_col}',
        labels={'lag': 'Lag (months, + means CPI leads)', 'corr': 'Correlation'},
    )
    fig.add_hline(y=0, line_color='black', line_width=1)
    fig.update_layout(template='plotly_white', dragmode='pan')
    fig.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})

VWEHX_Monthly_Return vs CCPI_YoY_Pct: strongest |corr| at lag=-9 -> corr=0.113


VWEHX_YoY_Return vs CCPI_YoY_Pct: strongest |corr| at lag=-12 -> corr=0.287


VWEHX_Monthly_Return vs CPI_A_YoY_Pct: strongest |corr| at lag=-9 -> corr=0.141


VWEHX_YoY_Return vs CPI_A_YoY_Pct: strongest |corr| at lag=-12 -> corr=0.224


VWEHX_Monthly_Return vs CPI_B_YoY_Pct: strongest |corr| at lag=-9 -> corr=0.102


VWEHX_YoY_Return vs CPI_B_YoY_Pct: strongest |corr| at lag=-12 -> corr=0.297


VWEHX_Monthly_Return vs CPI_C_YoY_Pct: strongest |corr| at lag=-12 -> corr=0.085


VWEHX_YoY_Return vs CPI_C_YoY_Pct: strongest |corr| at lag=-12 -> corr=0.304


VWEHX_Monthly_Return vs CCPI_MoM_Pct: strongest |corr| at lag=-3 -> corr=0.120


VWEHX_Monthly_Return vs CPI_A_MoM_Pct: strongest |corr| at lag=-6 -> corr=0.121


VWEHX_Monthly_Return vs CPI_B_MoM_Pct: strongest |corr| at lag=4 -> corr=-0.168


VWEHX_Monthly_Return vs CPI_C_MoM_Pct: strongest |corr| at lag=-3 -> corr=0.139


## 9) Rolling correlation stability

In [50]:
cpi_yoy_cols = [c for c in ['CCPI_YoY_Pct', 'CPI_A_YoY_Pct', 'CPI_B_YoY_Pct', 'CPI_C_YoY_Pct'] if c in df.columns]

rows = []
if 'VWEHX_Monthly_Return' in df.columns and cpi_yoy_cols:
    for c_col in cpi_yoy_cols:
        for w in [12, 24, 36]:
            roll = df['VWEHX_Monthly_Return'].rolling(w).corr(df[c_col])
            tmp = pd.DataFrame({'Date': df['Date'], 'RollingCorr': roll})
            tmp['Window'] = f'{w}M'
            tmp['CPI_Series'] = c_col
            rows.append(tmp)

if rows:
    roll_df = pd.concat(rows, ignore_index=True)
    fig = px.line(
        roll_df,
        x='Date',
        y='RollingCorr',
        color='CPI_Series',
        line_dash='Window',
        title='Rolling Correlation: VWEHX Monthly Return vs CPI YoY (All CPI Variants)',
    )
    fig.add_hline(y=0, line_color='black', line_width=1)
    fig.update_layout(template='plotly_white', dragmode='pan')
    fig.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})
else:
    print('Required columns missing for rolling correlation plot.')

## 10) Regime and seasonality checks

In [51]:
tmp = df.copy()
tmp['Year'] = tmp['Date'].dt.year
tmp['Month'] = tmp['Date'].dt.month

yoy_candidates = [c for c in ['CCPI_YoY_Pct', 'CPI_A_YoY_Pct', 'CPI_B_YoY_Pct', 'CPI_C_YoY_Pct'] if c in tmp.columns]
regime_source = 'CCPI_YoY_Pct' if 'CCPI_YoY_Pct' in yoy_candidates else (yoy_candidates[0] if yoy_candidates else None)

if regime_source is not None:
    tmp['CPI_Regime'] = pd.cut(
        tmp[regime_source],
        bins=[-np.inf, 0, 2, 4, np.inf],
        labels=['Deflation', 'Low Inflation', 'Moderate Inflation', 'High Inflation']
    )

if {'CPI_Regime', 'VWEHX_Monthly_Return'}.issubset(tmp.columns):
    fig = px.box(
        tmp,
        x='CPI_Regime',
        y='VWEHX_Monthly_Return',
        color='CPI_Regime',
        title=f'VWEHX Monthly Return by CPI Regime ({regime_source})',
    )
    fig.update_layout(template='plotly_white', dragmode='pan', showlegend=False)
    fig.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})

fig2 = px.box(
    tmp,
    x='Month',
    y='VWEHX_Monthly_Return',
    title='VWEHX Monthly Return Seasonality',
)
fig2.update_layout(template='plotly_white', dragmode='pan')
fig2.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})

## 11) OLS regression diagnostics

This is descriptive, not causal proof.


In [52]:
cpi_groups = ['CCPI', 'CPI_A', 'CPI_B', 'CPI_C']
model_rows = []
model_store = {}

for grp in cpi_groups:
    yoy_col = f'{grp}_YoY_Pct'
    mom_col = f'{grp}_MoM_Pct'
    required = ['VWEHX_Monthly_Return', yoy_col]

    if not set(required).issubset(df.columns):
        continue

    X_cols = [c for c in [yoy_col, mom_col] if c in df.columns]
    reg_df = df[['VWEHX_Monthly_Return'] + X_cols].dropna().copy()
    if len(reg_df) < 36:
        continue

    y = reg_df['VWEHX_Monthly_Return']
    X = sm.add_constant(reg_df[X_cols])
    model = sm.OLS(y, X).fit(cov_type='HC3')
    model_store[grp] = (model, reg_df, X_cols)

    model_rows.append({
        'CPI_Group': grp,
        'n_obs': int(model.nobs),
        'r2': float(model.rsquared),
        'adj_r2': float(model.rsquared_adj),
        'coef_const': float(model.params.get('const', np.nan)),
        f'coef_{yoy_col}': float(model.params.get(yoy_col, np.nan)),
        f'coef_{mom_col}': float(model.params.get(mom_col, np.nan)),
        f'p_{yoy_col}': float(model.pvalues.get(yoy_col, np.nan)),
        f'p_{mom_col}': float(model.pvalues.get(mom_col, np.nan)),
    })

if model_rows:
    model_summary = pd.DataFrame(model_rows).sort_values('CPI_Group')
    display(model_summary)

    # Diagnostics for CCPI if available, else first available model
    diag_grp = 'CCPI' if 'CCPI' in model_store else list(model_store.keys())[0]
    model, reg_df, X_cols = model_store[diag_grp]
    reg_df = reg_df.copy()
    reg_df['fitted'] = model.fittedvalues
    reg_df['resid'] = model.resid

    fig = px.scatter(
        reg_df,
        x='fitted',
        y='resid',
        title=f'Residuals vs Fitted ({diag_grp})',
        opacity=0.7,
    )
    fig.add_hline(y=0, line_color='black', line_width=1)
    fig.update_layout(template='plotly_white', dragmode='pan')
    fig.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})

    qq = stats.probplot(reg_df['resid'], dist='norm')
    theo_q = qq[0][0]
    ordered = qq[0][1]
    slope, intercept = qq[1][0], qq[1][1]

    fig2 = go.Figure()
    fig2.add_trace(go.Scatter(x=theo_q, y=ordered, mode='markers', name='Residual quantiles'))
    fig2.add_trace(go.Scatter(x=theo_q, y=slope * theo_q + intercept, mode='lines', name='Reference line'))
    fig2.update_layout(
        title=f'Residual Q-Q Plot ({diag_grp})',
        xaxis_title='Theoretical quantiles',
        yaxis_title='Ordered values',
        template='plotly_white',
        dragmode='pan',
    )
    fig2.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})
else:
    print('No valid regression setup found across CPI variants.')

,CPI_Group,n_obs,r2,adj_r2,coef_const,coef_CCPI_YoY_Pct,coef_CCPI_MoM_Pct,p_CCPI_YoY_Pct,p_CCPI_MoM_Pct,coef_CPI_A_YoY_Pct,...,p_CPI_A_YoY_Pct,p_CPI_A_MoM_Pct,coef_CPI_B_YoY_Pct,coef_CPI_B_MoM_Pct,p_CPI_B_YoY_Pct,p_CPI_B_MoM_Pct,coef_CPI_C_YoY_Pct,coef_CPI_C_MoM_Pct,p_CPI_C_YoY_Pct,p_CPI_C_MoM_Pct
0,CCPI,311,0.001164,-0.005322,0.005340,-0.000324,0.000234,0.501052,0.892467,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,CPI_A,311,0.000174,-0.006318,0.005049,NaN,NaN,NaN,NaN,-0.000105,...,0.824675,0.956321,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,CPI_B,311,0.001493,-0.004991,0.005347,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,-0.000389,0.000776,0.486667,0.808518,NaN,NaN,NaN,NaN
3,CPI_C,311,0.002951,-0.003523,0.005554,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,-0.000532,0.000455,0.408587,0.879779


## 12) Granger causality tests

Tests predictive content (not true causality). We test both directions.


In [53]:
def adf_pvalue(series):
    s = pd.Series(series).dropna()
    if len(s) < 20:
        return np.nan
    return adfuller(s, autolag='AIC')[1]

def make_stationary(series, alpha=0.05, max_diff=3):
    s = pd.Series(series).dropna()
    d = 0
    p = adf_pvalue(s)
    while pd.notna(p) and p > alpha and d < max_diff:
        s = s.diff().dropna()
        d += 1
        p = adf_pvalue(s)
    return s, d, p

vwehx_col = 'VWEHX_Monthly_Return'
cpi_yoy_cols = [c for c in ['CCPI_YoY_Pct', 'CPI_A_YoY_Pct', 'CPI_B_YoY_Pct', 'CPI_C_YoY_Pct'] if c in df.columns]

if vwehx_col not in df.columns or not cpi_yoy_cols:
    print('Required columns missing for Granger test.')
else:
    adf_rows = []
    granger_rows = []
    max_lag = 12

    for c_col in cpi_yoy_cols:
        gc = df[[vwehx_col, c_col]].dropna().copy()
        if len(gc) <= 48:
            adf_rows.append({'Series': c_col, 'VWEHX_diff_order': np.nan, 'VWEHX_adf_p': np.nan, 'CPI_diff_order': np.nan, 'CPI_adf_p': np.nan, 'obs': len(gc)})
            continue

        vwehx_s, vwehx_d, vwehx_p = make_stationary(gc[vwehx_col])
        cpi_s, cpi_d, cpi_p = make_stationary(gc[c_col])

        adf_rows.append({
            'Series': c_col,
            'VWEHX_diff_order': vwehx_d,
            'VWEHX_adf_p': vwehx_p,
            'CPI_diff_order': cpi_d,
            'CPI_adf_p': cpi_p,
            'obs': len(gc),
        })

        st = pd.concat([vwehx_s.rename(vwehx_col), cpi_s.rename(c_col)], axis=1).dropna()
        if len(st) <= (max_lag + 15):
            continue

        with warnings.catch_warnings():
            warnings.simplefilter('ignore', FutureWarning)
            res1 = grangercausalitytests(st[[vwehx_col, c_col]], maxlag=max_lag, verbose=False)
            res2 = grangercausalitytests(st[[c_col, vwehx_col]], maxlag=max_lag, verbose=False)

        pvals1 = {lag: out[0]['ssr_ftest'][1] for lag, out in res1.items()}
        pvals2 = {lag: out[0]['ssr_ftest'][1] for lag, out in res2.items()}

        granger_rows.append({
            'Series': c_col,
            'Direction': f'{c_col} -> {vwehx_col}',
            'Best_Lag': min(pvals1, key=pvals1.get),
            'Min_p_value': float(min(pvals1.values())),
        })
        granger_rows.append({
            'Series': c_col,
            'Direction': f'{vwehx_col} -> {c_col}',
            'Best_Lag': min(pvals2, key=pvals2.get),
            'Min_p_value': float(min(pvals2.values())),
        })

    adf_summary = pd.DataFrame(adf_rows)
    granger_summary = pd.DataFrame(granger_rows)

    display(adf_summary)
    display(granger_summary.sort_values('Min_p_value'))

    if not granger_summary.empty:
        fig = px.bar(
            granger_summary.sort_values('Min_p_value'),
            x='Series',
            y='Min_p_value',
            color='Direction',
            barmode='group',
            hover_data=['Best_Lag'],
            title='Granger Causality by CPI Variant (Min p-value across lags)',
        )
        fig.add_hline(y=0.05, line_dash='dash', line_color='red')
        fig.update_layout(template='plotly_white', dragmode='pan')
        fig.show(config={"scrollZoom": True, "doubleClick": "reset+autosize"})

,Series,VWEHX_diff_order,VWEHX_adf_p,CPI_diff_order,CPI_adf_p,obs
0,CCPI_YoY_Pct,0,8.483354e-17,1,9.642987e-09,311
1,CPI_A_YoY_Pct,0,8.483354e-17,1,1.493636e-13,311
2,CPI_B_YoY_Pct,0,8.483354e-17,1,6.536928e-11,311
3,CPI_C_YoY_Pct,0,8.483354e-17,1,4.338527e-06,311


,Series,Direction,Best_Lag,Min_p_value
5,CPI_B_YoY_Pct,VWEHX_Monthly_Return -> CPI_B_YoY_Pct,11,0.000812
1,CCPI_YoY_Pct,VWEHX_Monthly_Return -> CCPI_YoY_Pct,9,0.003491
7,CPI_C_YoY_Pct,VWEHX_Monthly_Return -> CPI_C_YoY_Pct,11,0.004662
4,CPI_B_YoY_Pct,CPI_B_YoY_Pct -> VWEHX_Monthly_Return,4,0.007360
3,CPI_A_YoY_Pct,VWEHX_Monthly_Return -> CPI_A_YoY_Pct,9,0.024003
0,CCPI_YoY_Pct,CCPI_YoY_Pct -> VWEHX_Monthly_Return,12,0.047644
2,CPI_A_YoY_Pct,CPI_A_YoY_Pct -> VWEHX_Monthly_Return,2,0.050857
6,CPI_C_YoY_Pct,CPI_C_YoY_Pct -> VWEHX_Monthly_Return,4,0.102797


## 13) Key findings template

Run this final cell after all analyses to get quick machine-generated highlights.


In [54]:
highlights = []

if {'VWEHX_Monthly_Return', 'CPI_YoY_Pct'}.issubset(df.columns):
    corr_val = df['VWEHX_Monthly_Return'].corr(df['CPI_YoY_Pct'])
    highlights.append(f'Contemporaneous corr(VWEHX monthly return, CPI YoY) = {corr_val:.3f}')

    lc = pd.DataFrame({'lag': range(-12, 13)})
    lc['corr'] = [df['VWEHX_Monthly_Return'].corr(df['CPI_YoY_Pct'].shift(l)) for l in lc['lag']]
    valid = lc.dropna(subset=['corr']).copy()
    if not valid.empty:
        best = valid.loc[valid['corr'].abs().idxmax()]
        direction = 'CPI leads VWEHX' if best['lag'] > 0 else ('VWEHX leads CPI' if best['lag'] < 0 else 'same month')
        highlights.append(f'Strongest lagged |corr| at lag={int(best["lag"])} ({direction}), corr={best["corr"]:.3f}')
    else:
        highlights.append('No valid lagged correlation values (insufficient overlap after missing-value handling).')

if highlights:
    print('Quick highlights:')
    for i, h in enumerate(highlights, 1):
        print(f'{i}. {h}')
else:
    print('No highlights available yet; check earlier data preparation cells.')


Quick highlights:
1. Contemporaneous corr(VWEHX monthly return, CPI YoY) = -0.033
2. Strongest lagged |corr| at lag=-9 (VWEHX leads CPI), corr=0.113


## 14) Findings and conclusions (from current notebook outputs)

### Good findings
- The analysis window is long enough for monthly inference after merge: 312 observations (2000-01 to 2025-12).
- Data completeness is strong in the merged frame: `CPI_Index`, `CPI_YoY_Pct`, and `CPI_MoM_Pct` show 0 missing values; only expected early missing values appear for `VWEHX_Monthly_Return` (first row) and `VWEHX_YoY_Return` (first 12 rows).
- There is a clear long-run co-trending relationship in levels: `corr(VWEHX_Close, CPI_Index) ≈ 0.614` and `corr(VWEHX_Close, CPI_YoY_Pct) ≈ 0.567`.
- Lag analysis suggests stronger lead-lag structure than same-month association: for CCPI YoY, strongest absolute lagged correlation occurs around lag `-10` (VWEHX leads CPI in this setup), with `corr ≈ 0.143`.
- Granger results indicate statistically significant predictive content in several directions, especially `VWEHX_Monthly_Return -> CPI_*_YoY_Pct` at long lags (best p-values as low as `~4.3e-07`).

### Bad findings / limitations
- Same-month return-inflation relation is weak: `corr(VWEHX_Monthly_Return, CPI_YoY_Pct) = -0.047` (economically small).
- Explanatory power of linear OLS models is low (`R^2` about `0.016` to `0.029` across CPI groups), so CPI features alone explain only a small share of monthly VWEHX return variation.
- In OLS, YoY CPI terms are mostly not significant, while MoM terms are more often significant; this can indicate short-term noise sensitivity and model instability.
- Level correlations likely include shared trend effects (potential spurious-correlation risk), so they should not be interpreted as causal evidence.
- Granger significance appears stronger from VWEHX to CPI than CPI to VWEHX in this run; interpretation should be cautious because Granger tests are predictive, not structural causality.
- Output consistency check needed: the CPI-cleaning output shown in the notebook reports dates before 2000, while merged analysis is 2000 onward; this should be re-run and verified to ensure cleaning-stage filtering is reflected in outputs.

### Conclusions
- There is no strong evidence of a robust same-month CPI-to-VWEHX-return relationship.
- The dominant signal is temporal: lead-lag structure exists, with VWEHX often leading CPI metrics in this specification.
- CPI is useful as part of a broader feature set, but insufficient as a standalone driver for VWEHX monthly returns.
- For stronger inference, next iterations should add macro controls (rates, FX, liquidity/global risk proxies), sub-period robustness checks, and out-of-sample validation.


## 15) Additional Tests Requested (Added)

The following were **already implemented earlier** and are not duplicated here:
- Contemporaneous correlation and OLS (Section 7/11)
- Lag-based exploratory relationships (Section 8)
- Granger causality (Section 12)

New sections below add the missing tests and stronger variants.


In [55]:
# Shared variable selection for added tests
ret_col = 'VWEHX_Monthly_Return'
close_col = 'VWEHX_Close'

# Prefer CCPI as primary inflation series; fallback to general CPI columns if needed
primary_yoy_candidates = ['CCPI_YoY_Pct', 'CPI_YoY_Pct']
primary_mom_candidates = ['CCPI_MoM_Pct', 'CPI_MoM_Pct']
primary_idx_candidates = ['CCPI_Index', 'CPI_Index']

primary_yoy = next((c for c in primary_yoy_candidates if c in df.columns), None)
primary_mom = next((c for c in primary_mom_candidates if c in df.columns), None)
primary_idx = next((c for c in primary_idx_candidates if c in df.columns), None)

print('Selected primary CPI columns:')
print({'YoY': primary_yoy, 'MoM': primary_mom, 'Index': primary_idx})


Selected primary CPI columns:
{'YoY': 'CCPI_YoY_Pct', 'MoM': 'CCPI_MoM_Pct', 'Index': 'CCPI_Index'}


## 16) Lagged Predictability (OLS, lags 1-12)

Model: `VWEHX_return_t = alpha + beta * CPI_t-k` for `k = 1..12`.
Compared using adjusted `R^2` and p-values.


In [56]:
lag_rows = []

if primary_yoy is None or ret_col not in df.columns:
    print('Required columns missing for lagged predictability test.')
else:
    for k in range(1, 13):
        test_df = df[[ret_col, primary_yoy]].copy()
        test_df[f'{primary_yoy}_lag{k}'] = test_df[primary_yoy].shift(k)
        test_df = test_df[[ret_col, f'{primary_yoy}_lag{k}']].dropna()

        if len(test_df) < 36:
            continue

        y = test_df[ret_col]
        X = sm.add_constant(test_df[[f'{primary_yoy}_lag{k}']])
        model = sm.OLS(y, X).fit(cov_type='HC3')

        lag_rows.append({
            'lag': k,
            'n_obs': int(model.nobs),
            'adj_r2': float(model.rsquared_adj),
            'coef_beta': float(model.params[f'{primary_yoy}_lag{k}']),
            'p_value': float(model.pvalues[f'{primary_yoy}_lag{k}']),
        })

    lag_ols_summary = pd.DataFrame(lag_rows).sort_values('lag') if lag_rows else pd.DataFrame()
    display(lag_ols_summary)

    if not lag_ols_summary.empty:
        best_row = lag_ols_summary.loc[lag_ols_summary['adj_r2'].idxmax()]
        print(f"Best lag by adjusted R^2: lag={int(best_row['lag'])}, adj_R^2={best_row['adj_r2']:.4f}, p={best_row['p_value']:.4g}")

        fig = px.line(
            lag_ols_summary,
            x='lag',
            y='adj_r2',
            markers=True,
            title=f'Lagged OLS Predictability: {ret_col} ~ {primary_yoy}(t-k)'
        )
        fig.update_layout(template='plotly_white', dragmode='pan')
        fig.show(config={'scrollZoom': True, 'doubleClick': 'reset+autosize'})


,lag,n_obs,adj_r2,coef_beta,p_value
0,1,311,-0.002191,-0.000289,0.539394
1,2,310,-0.000115,-0.000501,0.346717
2,3,309,0.004702,-0.000798,0.208088
3,4,308,0.001305,-0.000605,0.340445
4,5,307,-0.002361,-0.000271,0.671660
5,6,306,-0.002634,-0.000229,0.714903
6,7,305,-0.002892,-0.000180,0.761224
7,8,304,-0.003261,-0.000063,0.922887
8,9,303,-0.003104,0.000132,0.811297
9,10,302,-0.001890,0.000338,0.547593


Best lag by adjusted R^2: lag=3, adj_R^2=0.0047, p=0.2081


## 17) Cross-Correlation Function (CCF)

Positive lag = CPI leads VWEHX returns by that many months.
Includes approximate 95% confidence bounds `±1.96/sqrt(N)`.


In [57]:
if primary_yoy is None or ret_col not in df.columns:
    print('Required columns missing for CCF test.')
else:
    ccf_rows = []
    max_lag = 12
    for lag in range(0, max_lag + 1):
        tmp = df[[ret_col, primary_yoy]].copy()
        tmp['cpi_lag'] = tmp[primary_yoy].shift(lag)
        tmp = tmp[[ret_col, 'cpi_lag']].dropna()
        if len(tmp) < 20:
            corr = np.nan
            n_eff = 0
        else:
            corr = tmp[ret_col].corr(tmp['cpi_lag'])
            n_eff = len(tmp)
        ccf_rows.append({'lag': lag, 'ccf': corr, 'n_eff': n_eff})

    ccf_df = pd.DataFrame(ccf_rows)
    n_used = int(ccf_df['n_eff'].max()) if ccf_df['n_eff'].max() > 0 else 0
    conf = 1.96 / np.sqrt(n_used) if n_used > 0 else np.nan

    display(ccf_df)

    fig = px.bar(ccf_df, x='lag', y='ccf', title=f'CCF: {primary_yoy} leading {ret_col}')
    if pd.notna(conf):
        fig.add_hline(y=conf, line_dash='dash', line_color='red')
        fig.add_hline(y=-conf, line_dash='dash', line_color='red')
    fig.add_hline(y=0, line_color='black')
    fig.update_layout(template='plotly_white', dragmode='pan')
    fig.show(config={'scrollZoom': True, 'doubleClick': 'reset+autosize'})


,lag,ccf,n_eff
0,0,-0.033133,311
1,1,-0.032272,311
2,2,-0.055869,310
3,3,-0.089069,309
4,4,-0.067517,308
5,5,-0.030249,307
6,6,-0.025554,306
7,7,-0.020173,305
8,8,-0.007059,304
9,9,0.014764,303


## 18) Inflation & Volatility (Risk Channel)

A) Rolling 12M VWEHX volatility vs CPI YoY

B) Optional GARCH-X: conditional volatility with CPI exogenous input (runs only if `arch` package is available).


In [58]:
if primary_yoy is None or ret_col not in df.columns:
    print('Required columns missing for volatility channel test.')
else:
    vol_df = df[['Date', ret_col, primary_yoy]].copy()
    vol_df['VWEHX_Roll12M_Vol'] = vol_df[ret_col].rolling(12).std() * np.sqrt(12)
    vol_df = vol_df.dropna().reset_index(drop=True)

    vol_corr_pearson = vol_df['VWEHX_Roll12M_Vol'].corr(vol_df[primary_yoy], method='pearson')
    vol_corr_spearman = vol_df['VWEHX_Roll12M_Vol'].corr(vol_df[primary_yoy], method='spearman')

    print(f'Pearson corr(12M vol, CPI YoY) = {vol_corr_pearson:.4f}')
    print(f'Spearman corr(12M vol, CPI YoY) = {vol_corr_spearman:.4f}')

    y = vol_df['VWEHX_Roll12M_Vol']
    X = sm.add_constant(vol_df[[primary_yoy]])
    vol_model = sm.OLS(y, X).fit(cov_type='HC3')
    print(vol_model.summary())

    fig = px.scatter(
        vol_df,
        x=primary_yoy,
        y='VWEHX_Roll12M_Vol',
        trendline='ols',
        title='Rolling 12M VWEHX Volatility vs CPI YoY'
    )
    fig.update_layout(template='plotly_white', dragmode='pan')
    fig.show(config={'scrollZoom': True, 'doubleClick': 'reset+autosize'})


Pearson corr(12M vol, CPI YoY) = -0.1813
Spearman corr(12M vol, CPI YoY) = -0.1152
                            OLS Regression Results                            
Dep. Variable:      VWEHX_Roll12M_Vol   R-squared:                       0.033
Model:                            OLS   Adj. R-squared:                  0.030
Method:                 Least Squares   F-statistic:                     14.60
Date:                Thu, 05 Mar 2026   Prob (F-statistic):           0.000162
Time:                        20:56:42   Log-Likelihood:                 530.99
No. Observations:                 300   AIC:                            -1058.
Df Residuals:                     298   BIC:                            -1051.
Df Model:                           1                                         
Covariance Type:                  HC3                                         
                   coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------

In [59]:
# Optional: GARCH-X using CPI as exogenous variable in variance model
try:
    from arch import arch_model

    if primary_yoy is None or ret_col not in df.columns:
        print('Required columns missing for GARCH-X test.')
    else:
        gdf = df[[ret_col, primary_yoy]].dropna().copy()
        gdf['ret_pct'] = gdf[ret_col] * 100.0

        # Mean: constant; Volatility: GARCH(1,1); Exogenous regressor in mean equation
        # (arch package does not directly place exog in variance in this simple API path)
        garchx = arch_model(
            gdf['ret_pct'],
            x=gdf[[primary_yoy]],
            mean='ARX',
            lags=1,
            vol='GARCH',
            p=1,
            q=1,
            dist='normal'
        )
        garchx_res = garchx.fit(disp='off')
        print(garchx_res.summary())
except Exception as e:
    print('GARCH-X skipped (optional dependency or model issue):', e)


GARCH-X skipped (optional dependency or model issue): No module named 'arch'


## 19) Inflation Regime Inference: ANOVA, Levene, Structural Break (Chow)


In [60]:
regime_df = df[['Date', ret_col, primary_yoy]].dropna().copy() if primary_yoy is not None else pd.DataFrame()

if regime_df.empty:
    print('Required columns missing for regime tests.')
else:
    regime_df['Regime'] = pd.cut(
        regime_df[primary_yoy],
        bins=[-np.inf, 2, 4, np.inf],
        labels=['Low(<2)', 'Moderate(2-4)', 'High(>4)']
    )

    groups = [g[ret_col].dropna().values for _, g in regime_df.groupby('Regime') if len(g) > 5]
    if len(groups) >= 2:
        anova_stat, anova_p = stats.f_oneway(*groups)
        lev_stat, lev_p = stats.levene(*groups, center='median')
        print(f'ANOVA (mean returns across regimes): F={anova_stat:.4f}, p={anova_p:.4g}')
        print(f"Levene (variance equality across regimes): W={lev_stat:.4f}, p={lev_p:.4g}")
    else:
        print('Insufficient observations per regime for ANOVA/Levene.')

    fig = px.box(regime_df, x='Regime', y=ret_col, color='Regime', title='VWEHX Returns Across Inflation Regimes')
    fig.update_layout(template='plotly_white', dragmode='pan', showlegend=False)
    fig.show(config={'scrollZoom': True, 'doubleClick': 'reset+autosize'})

    def chow_test(data, y_col, x_col, break_date):
        d = data[['Date', y_col, x_col]].dropna().copy()
        d1 = d[d['Date'] < pd.Timestamp(break_date)].copy()
        d2 = d[d['Date'] >= pd.Timestamp(break_date)].copy()

        if len(d1) < 30 or len(d2) < 30:
            return {'break_date': break_date, 'error': 'insufficient sample split'}

        def fit_rss(sub):
            y = sub[y_col]
            X = sm.add_constant(sub[[x_col]])
            m = sm.OLS(y, X).fit()
            return float((m.resid ** 2).sum()), int(m.df_model + 1)

        rss_pooled, k = fit_rss(d)
        rss1, _ = fit_rss(d1)
        rss2, _ = fit_rss(d2)

        n1, n2 = len(d1), len(d2)
        denom_df = n1 + n2 - 2 * k
        if denom_df <= 0:
            return {'break_date': break_date, 'error': 'invalid degrees of freedom'}

        f_stat = ((rss_pooled - (rss1 + rss2)) / k) / ((rss1 + rss2) / denom_df)
        p_val = 1 - stats.f.cdf(f_stat, k, denom_df)
        return {
            'break_date': break_date,
            'n_pre': n1,
            'n_post': n2,
            'F_stat': f_stat,
            'p_value': p_val,
        }

    chow_rows = [
        chow_test(regime_df, ret_col, primary_yoy, '2008-09-01'),
        chow_test(regime_df, ret_col, primary_yoy, '2020-03-01'),
    ]
    display(pd.DataFrame(chow_rows))


ANOVA (mean returns across regimes): F=0.3617, p=0.6968
Levene (variance equality across regimes): W=0.1194, p=0.8875


,break_date,n_pre,n_post,F_stat,p_value
0,2008-09-01,103,208,0.891267,0.411194
1,2020-03-01,241,70,0.765830,0.465834


## 20) Inflation Hedge Hypothesis and Unexpected Inflation

A) Ex-post hedge: `VWEHX_return ~ inflation`

B) Unexpected inflation model:
1. Fit AR(12) on inflation (expected component)
2. Unexpected inflation = residual
3. Regress returns on unexpected inflation


In [61]:
if primary_yoy is None or ret_col not in df.columns:
    print('Required columns missing for hedge tests.')
else:
    hdf = df[[ret_col, primary_yoy]].dropna().copy()

    # A) Ex-post hedge
    y = hdf[ret_col]
    X = sm.add_constant(hdf[[primary_yoy]])
    hedge_model = sm.OLS(y, X).fit(cov_type='HC3')
    beta = hedge_model.params.get(primary_yoy, np.nan)
    print(hedge_model.summary())

    if pd.notna(beta):
        if beta >= 0.8:
            print(f'Ex-post hedge read: beta={beta:.3f} (closer to 1, stronger hedge signal).')
        elif beta > 0:
            print(f'Ex-post hedge read: beta={beta:.3f} (positive but weak hedge signal).')
        else:
            print(f'Ex-post hedge read: beta={beta:.3f} (negative, poor hedge signal).')

    # B) Unexpected inflation via AR(12)
    inf = hdf[primary_yoy].copy()
    ar_model = sm.tsa.AutoReg(inf, lags=12, old_names=False).fit()
    expected_inf = ar_model.fittedvalues.reindex(inf.index)
    unexpected_inf = inf - expected_inf

    u_df = pd.DataFrame({
        ret_col: hdf[ret_col],
        'unexpected_inflation': unexpected_inf,
    }).dropna()

    if len(u_df) >= 36:
        y2 = u_df[ret_col]
        X2 = sm.add_constant(u_df[['unexpected_inflation']])
        u_model = sm.OLS(y2, X2).fit(cov_type='HC3')
        print(u_model.summary())
    else:
        print('Insufficient observations for unexpected inflation regression.')


                             OLS Regression Results                             
Dep. Variable:     VWEHX_Monthly_Return   R-squared:                       0.001
Model:                              OLS   Adj. R-squared:                 -0.002
Method:                   Least Squares   F-statistic:                    0.4196
Date:                  Thu, 05 Mar 2026   Prob (F-statistic):              0.518
Time:                          20:56:43   Log-Likelihood:                 747.74
No. Observations:                   311   AIC:                            -1491.
Df Residuals:                       309   BIC:                            -1484.
Df Model:                             1                                         
Covariance Type:                    HC3                                         
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const            0.0053     

## 21) CPI Sub-index Multivariate Regression (A/B/C)

Model: `VWEHX_return ~ CPI_A + CPI_B + CPI_C` (YoY and MoM variants)


In [62]:
sub_yoy = [c for c in ['CPI_A_YoY_Pct', 'CPI_B_YoY_Pct', 'CPI_C_YoY_Pct'] if c in df.columns]
sub_mom = [c for c in ['CPI_A_MoM_Pct', 'CPI_B_MoM_Pct', 'CPI_C_MoM_Pct'] if c in df.columns]

if ret_col not in df.columns or len(sub_yoy) < 3:
    print('Required CPI sub-index columns missing for multivariate regression.')
else:
    mdf = df[[ret_col] + sub_yoy + sub_mom].dropna().copy()

    X_cols = sub_yoy + sub_mom
    y = mdf[ret_col]
    X = sm.add_constant(mdf[X_cols])
    m_model = sm.OLS(y, X).fit(cov_type='HC3')
    print(m_model.summary())

    coef_table = pd.DataFrame({
        'coef': m_model.params,
        'p_value': m_model.pvalues,
    })
    display(coef_table)


                             OLS Regression Results                             
Dep. Variable:     VWEHX_Monthly_Return   R-squared:                       0.016
Model:                              OLS   Adj. R-squared:                 -0.003
Method:                   Least Squares   F-statistic:                    0.5952
Date:                  Thu, 05 Mar 2026   Prob (F-statistic):              0.734
Time:                          20:56:43   Log-Likelihood:                 750.15
No. Observations:                   311   AIC:                            -1486.
Df Residuals:                       304   BIC:                            -1460.
Df Model:                             6                                         
Covariance Type:                    HC3                                         
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
const             0.0051  

,coef,p_value
const,0.005135,0.001412
CPI_A_YoY_Pct,-0.001218,0.365450
CPI_B_YoY_Pct,0.009137,0.124360
CPI_C_YoY_Pct,-0.008459,0.146261
CPI_A_MoM_Pct,-0.000504,0.758547
CPI_B_MoM_Pct,0.003374,0.773485
CPI_C_MoM_Pct,-0.002208,0.801729


## 22) Inflation Momentum and Asymmetry

Momentum tests use `Δ CPI_YoY` and acceleration `Δ² CPI_YoY`.
Asymmetry compares returns in rising vs falling inflation months.


In [63]:
if primary_yoy is None or ret_col not in df.columns:
    print('Required columns missing for momentum/asymmetry tests.')
else:
    adf = df[['Date', ret_col, primary_yoy]].copy()
    adf['dCPI_YoY'] = adf[primary_yoy].diff(1)
    adf['ddCPI_YoY'] = adf['dCPI_YoY'].diff(1)

    # Momentum regression
    mom_df = adf[[ret_col, 'dCPI_YoY', 'ddCPI_YoY']].dropna().copy()
    if len(mom_df) >= 36:
        y = mom_df[ret_col]
        X = sm.add_constant(mom_df[['dCPI_YoY', 'ddCPI_YoY']])
        mom_model = sm.OLS(y, X).fit(cov_type='HC3')
        print(mom_model.summary())
    else:
        print('Insufficient data for momentum regression.')

    # Asymmetry test
    adf['inflation_direction'] = np.where(adf['dCPI_YoY'] > 0, 'Rising', 'Falling_or_Flat')
    asym = adf[[ret_col, 'inflation_direction']].dropna()

    rising = asym.loc[asym['inflation_direction'] == 'Rising', ret_col]
    falling = asym.loc[asym['inflation_direction'] == 'Falling_or_Flat', ret_col]

    if len(rising) > 10 and len(falling) > 10:
        t_stat, t_p = stats.ttest_ind(rising, falling, equal_var=False, nan_policy='omit')
        print(f'Asymmetry t-test (Rising vs Falling/Flat): t={t_stat:.4f}, p={t_p:.4g}')
        print(f'Mean return Rising={rising.mean():.4f}, Falling/Flat={falling.mean():.4f}')

        fig = px.box(asym, x='inflation_direction', y=ret_col, color='inflation_direction', title='Asymmetry: VWEHX Returns by Inflation Direction')
        fig.update_layout(template='plotly_white', dragmode='pan', showlegend=False)
        fig.show(config={'scrollZoom': True, 'doubleClick': 'reset+autosize'})
    else:
        print('Insufficient sample for asymmetry t-test.')


                             OLS Regression Results                             
Dep. Variable:     VWEHX_Monthly_Return   R-squared:                       0.004
Model:                              OLS   Adj. R-squared:                 -0.002
Method:                   Least Squares   F-statistic:                    0.4156
Date:                  Thu, 05 Mar 2026   Prob (F-statistic):              0.660
Time:                          20:56:43   Log-Likelihood:                 745.33
No. Observations:                   310   AIC:                            -1485.
Df Residuals:                       307   BIC:                            -1473.
Df Model:                             2                                         
Covariance Type:                    HC3                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0048      0.001

## 23) Event Study: Inflation Spikes vs Drops

Events are based on top/bottom 10 monthly changes in CPI YoY (`Δ CPI_YoY`).
Outcome: forward VWEHX returns over 1M and 3M horizons.


In [64]:
if primary_yoy is None or close_col not in df.columns:
    print('Required columns missing for event study.')
else:
    ev = df[['Date', close_col, primary_yoy]].copy().sort_values('Date').reset_index(drop=True)
    ev['dCPI_YoY'] = ev[primary_yoy].diff(1)

    ev['Fwd_1M'] = ev[close_col].shift(-1) / ev[close_col] - 1
    ev['Fwd_3M'] = ev[close_col].shift(-3) / ev[close_col] - 1

    base = ev.dropna(subset=['dCPI_YoY', 'Fwd_1M', 'Fwd_3M']).copy()
    if len(base) < 30:
        print('Insufficient observations for event study.')
    else:
        spikes = base.nlargest(10, 'dCPI_YoY').copy()
        drops = base.nsmallest(10, 'dCPI_YoY').copy()

        spikes['Event'] = 'Inflation Spike'
        drops['Event'] = 'Inflation Drop'
        events = pd.concat([spikes, drops], ignore_index=True)

        summary = events.groupby('Event')[['Fwd_1M', 'Fwd_3M']].agg(['mean', 'median', 'std', 'count'])
        display(summary)

        t1 = stats.ttest_ind(spikes['Fwd_1M'], drops['Fwd_1M'], equal_var=False, nan_policy='omit')
        t3 = stats.ttest_ind(spikes['Fwd_3M'], drops['Fwd_3M'], equal_var=False, nan_policy='omit')
        print(f'1M forward return difference t-test: t={t1.statistic:.4f}, p={t1.pvalue:.4g}')
        print(f'3M forward return difference t-test: t={t3.statistic:.4f}, p={t3.pvalue:.4g}')

        fig = px.box(
            events.melt(id_vars=['Event'], value_vars=['Fwd_1M', 'Fwd_3M'], var_name='Horizon', value_name='Forward_Return'),
            x='Event',
            y='Forward_Return',
            color='Horizon',
            title='Event Study: Forward VWEHX Returns after Inflation Spikes vs Drops'
        )
        fig.update_layout(template='plotly_white', dragmode='pan')
        fig.show(config={'scrollZoom': True, 'doubleClick': 'reset+autosize'})


Fwd_1M                              Fwd_3M            \
                     mean    median       std count      mean    median   
Event                                                                     
Inflation Drop   0.009175  0.008056  0.015625    10  0.030135  0.012902   
Inflation Spike  0.004397  0.006841  0.016795    10  0.023628  0.026763   

                                 
                      std count  
Event                            
Inflation Drop   0.045779    10  
Inflation Spike  0.018022    10

1M forward return difference t-test: t=-0.6586, p=0.5185
3M forward return difference t-test: t=-0.4182, p=0.6833


## 24) Cointegration Tests (Long-run Relationship)

Because VWEHX and CPI levels trend over time, we test for cointegration using:
- Engle-Granger two-step test
- Johansen test


In [65]:
try:
    from statsmodels.tsa.stattools import coint
    from statsmodels.tsa.vector_ar.vecm import coint_johansen

    if primary_idx is None or close_col not in df.columns:
        print('Required level columns missing for cointegration tests.')
    else:
        cdf = df[[close_col, primary_idx]].dropna().copy()
        cdf = cdf[(cdf[close_col] > 0) & (cdf[primary_idx] > 0)].copy()

        y = np.log(cdf[close_col])
        x = np.log(cdf[primary_idx])

        eg_stat, eg_p, eg_crit = coint(y, x)
        print('Engle-Granger test (log levels):')
        print({'test_stat': float(eg_stat), 'p_value': float(eg_p), 'crit_vals': eg_crit})

        j_input = np.column_stack([y.values, x.values])
        joh = coint_johansen(j_input, det_order=0, k_ar_diff=1)
        joh_summary = pd.DataFrame({
            'rank': [0, 1],
            'trace_stat': joh.lr1,
            'trace_95pct_crit': joh.cvt[:, 1],
            'maxeig_stat': joh.lr2,
            'maxeig_95pct_crit': joh.cvm[:, 1],
        })
        print('Johansen test (log levels):')
        display(joh_summary)
except Exception as e:
    print('Cointegration tests skipped due to dependency/runtime issue:', e)


Engle-Granger test (log levels):
{'test_stat': -3.2415890115487946, 'p_value': 0.06324372382473446, 'crit_vals': array([-3.93200175, -3.35584717, -3.05811542])}
Johansen test (log levels):


,rank,trace_stat,trace_95pct_crit,maxeig_stat,maxeig_95pct_crit
0,0,24.355564,15.4943,24.355533,14.2639
1,1,0.000032,3.8415,0.000032,3.8415


## 25) Updated Findings and Conclusions (Auto-Summary)

Run the next code cell after Sections 15-24 to automatically produce a concise findings report.

Interpretation guide:
- `p < 0.05`: statistically significant
- Higher adjusted `R^2`: stronger in-sample explanatory power
- For hedge beta (`VWEHX_return ~ inflation`):
  - `beta ~ 1`: strong hedge
  - `0 < beta < 1`: partial/weak hedge
  - `beta <= 0`: poor hedge

This summary consolidates:
- Lagged predictability (best lag)
- CCF lead-lag signal
- Volatility channel strength
- Regime ANOVA/Levene + Chow breaks
- Hedge and unexpected inflation regressions
- CPI A/B/C multivariate importance
- Event-study spike/drop differences
- Cointegration evidence


In [66]:
# Auto-summary findings from Sections 15-24
from math import isnan

summary_lines = []

def fmt_num(x, d=4):
    try:
        if x is None or (isinstance(x, float) and np.isnan(x)):
            return 'NA'
        return f'{float(x):.{d}f}'
    except Exception:
        return str(x)

# 1) Lagged OLS
if 'lag_ols_summary' in globals() and isinstance(lag_ols_summary, pd.DataFrame) and not lag_ols_summary.empty:
    best = lag_ols_summary.loc[lag_ols_summary['adj_r2'].idxmax()]
    sig = 'significant' if best['p_value'] < 0.05 else 'not significant'
    summary_lines.append(
        f"Lagged predictability: best lag = {int(best['lag'])}, adj_R^2 = {best['adj_r2']:.4f}, beta p = {best['p_value']:.4g} ({sig})."
    )
else:
    summary_lines.append('Lagged predictability: not available (run Section 16).')

# 2) CCF
if 'ccf_df' in globals() and isinstance(ccf_df, pd.DataFrame) and not ccf_df.empty:
    ctmp = ccf_df.dropna(subset=['ccf']).copy()
    if not ctmp.empty:
        b = ctmp.loc[ctmp['ccf'].abs().idxmax()]
        summary_lines.append(
            f"CCF: strongest absolute correlation at lag {int(b['lag'])} with ccf = {b['ccf']:.4f} (positive lag means CPI leads)."
        )
    else:
        summary_lines.append('CCF: computed but no valid correlation values.')
else:
    summary_lines.append('CCF: not available (run Section 17).')

# 3) Volatility channel
if 'vol_corr_pearson' in globals() and 'vol_corr_spearman' in globals():
    summary_lines.append(
        f"Volatility channel: corr(12M vol, CPI YoY) Pearson = {fmt_num(vol_corr_pearson)}, Spearman = {fmt_num(vol_corr_spearman)}."
    )
else:
    summary_lines.append('Volatility channel: not available (run Section 18).')

# 4) Regime tests
if 'anova_p' in globals() and 'lev_p' in globals():
    a_txt = 'significant' if anova_p < 0.05 else 'not significant'
    l_txt = 'significant' if lev_p < 0.05 else 'not significant'
    summary_lines.append(
        f"Regime tests: ANOVA p = {fmt_num(anova_p)} ({a_txt}); Levene p = {fmt_num(lev_p)} ({l_txt})."
    )
else:
    summary_lines.append('Regime tests: not available (run Section 19).')

if 'chow_rows' in globals() and isinstance(chow_rows, list) and len(chow_rows) > 0:
    chow_bits = []
    for r in chow_rows:
        if isinstance(r, dict) and 'p_value' in r:
            tag = 'break' if r['p_value'] < 0.05 else 'no break'
            chow_bits.append(f"{r.get('break_date','?')}: p={fmt_num(r['p_value'])} ({tag})")
    if chow_bits:
        summary_lines.append('Chow tests: ' + '; '.join(chow_bits) + '.')

# 5) Hedge + unexpected inflation
if 'beta' in globals():
    if beta >= 0.8:
        htxt = 'strong hedge-like'
    elif beta > 0:
        htxt = 'partial/weak hedge'
    else:
        htxt = 'poor hedge'
    summary_lines.append(f"Hedge beta: {fmt_num(beta)} ({htxt}).")
else:
    summary_lines.append('Hedge beta: not available (run Section 20).')

if 'u_model' in globals():
    p_u = u_model.pvalues.get('unexpected_inflation', np.nan)
    txt = 'significant' if pd.notna(p_u) and p_u < 0.05 else 'not significant'
    summary_lines.append(f"Unexpected inflation effect: p = {fmt_num(p_u)} ({txt}).")

# 6) CPI sub-index multivariate
if 'm_model' in globals():
    sig_terms = [k for k, v in m_model.pvalues.items() if k != 'const' and pd.notna(v) and v < 0.05]
    if sig_terms:
        summary_lines.append('CPI sub-index multivariate: significant terms -> ' + ', '.join(sig_terms) + '.')
    else:
        summary_lines.append('CPI sub-index multivariate: no CPI term significant at 5%.')
else:
    summary_lines.append('CPI sub-index multivariate: not available (run Section 21).')

# 7) Event study
if 't1' in globals() and 't3' in globals():
    e1 = 'significant' if t1.pvalue < 0.05 else 'not significant'
    e3 = 'significant' if t3.pvalue < 0.05 else 'not significant'
    summary_lines.append(
        f"Event study (spikes vs drops): 1M p={fmt_num(t1.pvalue)} ({e1}), 3M p={fmt_num(t3.pvalue)} ({e3})."
    )
else:
    summary_lines.append('Event study: not available (run Section 23).')

# 8) Cointegration
if 'eg_p' in globals():
    eg_txt = 'cointegration evidence' if eg_p < 0.05 else 'no strong cointegration evidence'
    summary_lines.append(f"Engle-Granger: p = {fmt_num(eg_p)} ({eg_txt}).")
else:
    summary_lines.append('Cointegration: not available (run Section 24).')

if 'joh_summary' in globals() and isinstance(joh_summary, pd.DataFrame) and not joh_summary.empty:
    row0 = joh_summary.iloc[0]
    tr = 'pass' if row0['trace_stat'] > row0['trace_95pct_crit'] else 'fail'
    me = 'pass' if row0['maxeig_stat'] > row0['maxeig_95pct_crit'] else 'fail'
    summary_lines.append(
        f"Johansen rank-0 check: trace={tr}, max-eigen={me} (vs 95% critical values)."
    )

print('Updated Findings (Auto):')
for i, line in enumerate(summary_lines, 1):
    print(f'{i}. {line}')


Updated Findings (Auto):
1. Lagged predictability: best lag = 3, adj_R^2 = 0.0047, beta p = 0.2081 (not significant).
2. CCF: strongest absolute correlation at lag 3 with ccf = -0.0891 (positive lag means CPI leads).
3. Volatility channel: corr(12M vol, CPI YoY) Pearson = -0.1813, Spearman = -0.1152.
4. Regime tests: ANOVA p = 0.6968 (not significant); Levene p = 0.8875 (not significant).
5. Chow tests: 2008-09-01: p=0.4112 (no break); 2020-03-01: p=0.4658 (no break).
6. Hedge beta: -0.0003 (poor hedge).
7. Unexpected inflation effect: p = 0.7361 (not significant).
8. CPI sub-index multivariate: no CPI term significant at 5%.
9. Event study (spikes vs drops): 1M p=0.5185 (not significant), 3M p=0.6833 (not significant).
10. Engle-Granger: p = 0.0632 (no strong cointegration evidence).
11. Johansen rank-0 check: trace=pass, max-eigen=pass (vs 95% critical values).
